add enemy detection function

karamarimo opened the pull request:
this function checks whether an enemy sees the player considering
obstacles

Commit:
add enemy detection function

this function checks whether an enemy sees the player considering
obstacles

kameike commented on HachitaiGame.xcodeproj/project.pbxproj:
```diff
@@ -1,945 +1,440 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Xcode Project のマージはマージツールを使うといいよ
　https://github.com/simonwagner/mergepbx

kameike replied:
これは次回からだね

kameike commented on HachitaiGame/EnemyController.swift:
```diff
@@ -0,0 +1,60 @@
+import SKTUtils
+
+class EnemyController {
+    public var enemies: [EnemyNode] = []
```
enemy contorllerが enemyのことを知りすぎてるのが結構問題
enemyとplayerを密結合にしてしまっている

MVCモデルのcontrollerの意味なら 実質ViewはSKSpriteNodeになるのでEnemyに実装したほうがいいと思う

kameike commented on HachitaiGame/EnemyController.swift:
```diff
@@ -0,0 +1,60 @@
+    }
+    
+    private func enemyDetectPlayer(enemy: EnemyNode) -> Bool {
+        let vec = CGPointSubtract(player!.position, enemy.position)
```
これはEnemy内に detect(target: SKSpriteNode)で個別に実装するといいと思う。

kameike commented on HachitaiGame/EnemyController.swift:
```diff
@@ -0,0 +1,60 @@
+    public func update() {
+        for enemy in enemies {
+            if enemyDetectPlayer(enemy: enemy) {
+                enemy.color = UIColor.purple
```
ここでdetect(target: SKSpriteNode) にplayerを渡してあげる。

そうするとこのクラスの役目はviewのアップデートだからどちらかというとマネージャークラスだね。
detect(target: SKSpriteNode)は今のようにBoolを返せば処理できるはず。

ただし見つかった場合UI等いろんなところでイベントを処理することがあるから
detect(target: SKSpriteNode)はVoidを返すようにして
PUBSUBモデルを取り入れるとより疎結合を保つことができる
https://www.pubnub.com/ を使ったりしよう
podからインストールできる。

クラスのリネームもしてしまおう

kameike reviewed (changes requested):
レビュー遅くなってごめんねえ

Commit:
fix enemy and player

-renamed EnemyController to EnemyManager
-moved enemy detection function from EnemyManager to EnemyNode class
-fixed player’s animation speed not being updated in touchesBegan

karamarimo commented:
pubnubってネット越しに通信するみたいですが、ゲームシステムにも使えるんですか？pubnubのネットワークを介さずに通信できるんでしょうか？

kameike commented:
ごめん　https://github.com/cesarferreira/SwiftEventBus
　こっちだった（土下座）

Commit:
update enemy detection to use SwiftEventBus

-now EnemyNode notifies its detection using SwiftEventBus (pub-sub
style)
-GameScene now restarts when an enemy finds the player
-updated enemy animation to walk around
-player’s physics body is now circle-shaped

karamarimo commented on HachitaiGame/Enemy.swift:
```diff
@@ -42,11 +39,67 @@ class EnemyNode: SKSpriteNode {
+            self.run(SKAction.move(by: vec, duration: 3.0))
+        }
+        
+        patrolAnimation = SKAction.repeatForever(SKAction.sequence([
```
敵が歩き回るようになりました。

kameike replied:
:)

kameike replied:
ここの部分関数に分けたほうがいいかもね
private func attchAnimation みたいな

kameike commented on HachitaiGame/Enemy.swift:
```diff
@@ -42,11 +39,67 @@ class EnemyNode: SKSpriteNode {
-    
+
+    public func detect(target: SKSpriteNode) {
+        let detectingPlayer = { () -> Bool in 
```
ここなんでlambdaにしているの？

karamarimo replied:
if を３重にするより見栄えがいいかなと思ったので...functionを別に作ったほうがいいですか？

kameike replied:
いや、そのままλとっちゃって
unobstructed そのまま使えばよくない？

karamarimo replied:
あっそうですね。
targetが見えなかったら何も処理しなくていいってことを見落としてました。

kameike commented on HachitaiGame/Enemy.swift:
```diff
@@ -42,11 +39,67 @@ class EnemyNode: SKSpriteNode {
+            
+            // check obstacles
+            var unobstructed = true
+            self.scene?.physicsWorld.enumerateBodies(alongRayStart: self.position, end: target.position, using: { (body: SKPhysicsBody, point: CGPoint, normal: CGVector, stop: UnsafeMutablePointer<ObjCBool>) in
```
現在走査線を一本しか出していないみたいだけど、敵の視野の実装どうする感じなの？
イメージはどんな感じ？

kameike replied:
あとあとここ　enumerateBodiesの関数が長いから適当に改行で切り分けたほうがいいかもね

karamarimo replied:
敵の視野は、楽な方法を模索中です。lightnodeを使って表現するっていうのを考えましたが、lightnodeは全方位にしか光が出せないし、影の表現がチープ(光が障害物をはみ出す)なんで諦めました。
地道に計算するしかないんでしょうかね。

kameike replied:
多分地道にするしかないのではー

kameike commented on HachitaiGame/Enemy.swift:
```diff
@@ -8,28 +8,25 @@
+//        self.addChild(directionIndicator)
         
         let fovPath = CGMutablePath()
         fovPath.move(to: CGPoint(x: 0, y: 0))
```
ここも  pathを書くfunctionに分けたほうがいいかもね。

多分枚フレームごとに描画し直す感じじゃないのかな？

kameike replied:
ということで毎フレーム障害物に遮られる視野を描かないとね。

障害物の移動の後だから、magenerからのeventをListenしてmanger.getAllObstacleみたいな実装がよいのではないでしょうか？

kameike commented on HachitaiGame/Enemy.swift:
```diff
@@ -8,28 +8,25 @@
+    let walkDistance: CGFloat = 100.0
+    
+    var lookAroundAnimation: SKAction!
+    var patrolAnimation: SKAction!
```
変数はprivateにしよう

kameike commented on HachitaiGame/Enemy.swift:
```diff
@@ -42,11 +39,67 @@ class EnemyNode: SKSpriteNode {
     }
-    
+
+    public func detect(target: SKSpriteNode) {
```
ここのスコープはinternalが良いのではないでしょうか？

karamarimo replied:
あ、internalの存在を知りませんでした。そうですね。

kameike commented on HachitaiGame/EnemyManager.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    public func update() {
+        for enemy in enemies {
+            enemy.detect(target: player!)
```
ちなみにここせっかくだからeventBusで実装してもいいかもね。

karamarimo replied:
EnemyManagerがpublish、EnemyがsubscribeしてEnemyにdetectさせるってことですか？

kameike replied:
そうそう。

kameike commented on HachitaiGame/GameScene.swift:
```diff
@@ -44,31 +41,33 @@ class GameScene: SKScene {
+        enemyController = EnemyManager(enemies: [enemy], player: player)
+        
+        SwiftEventBus.onMainThread(self, name: "i see you") { (ntf: Notification!) in
+            self.restartScene()
```
i see you がハードコーディングになっているから、どっかにclassをつくってletなプロパティにしようか

kameike replied:
Game.Enemy.findTarget 見たいにかけるといいかもね。

kameike reviewed (changes requested):
コードレビューしたよ。いいねじっこうしたら楽しい

Commit:
fix enemy node and enemy manager

-enemy node’s per-frame functions are now called through event bus
-event name for event bus are now declared in GameConst

karamarimo commented:
scene が切り替わってもなぜか enemy node がメモリから消えません...。
参照のグラフ
![メモリグラフ](http://imgur.com/oBbmVSn.png)

kameike commented:
ちゃんとマネージャーとかは消えてるんだよね？

karamarimo commented:
あとなぜかPlayerControllerNodeが最初から2つメモリにあります。ゲームオーバーを繰り返しても増えません。
GameScene、PlayerNode、EnemyManagerとかはちゃんと消えます。

kameike commented on HachitaiGame/Enemy.swift:
```diff
@@ -8,45 +8,118 @@
+        }
+        
+        if unobstructed {
+            SwiftEventBus.post(GameConst.Enemy.playerFound, sender: self)
```
ここでEnemy.removeAllActionしたらメモリリークなるなるはず

karamarimo replied:
gameSceneFinishedイベントのとこでもいいですか？

kameike commented on HachitaiGame/GameConst.swift:
```diff
@@ -0,0 +1,20 @@
+        static let gameSceneFinished = "game scene finished"
+    }
+    struct Enemy {
+        static let playerFound = "player found"
```
これイベントの名前だから[Enemy.foundPlayer]的な文にしたほうがいいと思う。

kameike replied:
イベントは現在形がよいんだっけ、ちょっと調べて見てくだされ〜

kameike commented on HachitaiGame/Enemy.swift:
```diff
@@ -8,45 +8,118 @@
+        }
+        SwiftEventBus.onMainThread(self, name: GameConst.Game.gameSceneFinished) {
+            (ntf: Notification!) in
+            SwiftEventBus.unregister(self)
```
ここちゃんとunregisterしてて良いと思います。素晴らしい〜〜

kameike reviewed (changes requested):
でばっかで止めて見たらGameSceneじたい二回つくられてるね。原因は不明（おそらくStartScene viewloadedでシーンを切り替えちゃってるから）

でもPlayerControllerがリークしてるのは謎。ちなみにSKButtonめっちゃリークしてる笑

ラムダでキャプチャしちゃってるからねぇ

Commit:
add label node showing touch coordinates

-added a SKLabelNode that shows the last-touched point’s coordinates.
-enemy node now calls its removeAllAction when the scene finishes
-added some functions to Utility class for calculating a path for enemy
vision indicator
-EnemyManager now passes all obstacle objects when posting updateVision
event

Commit:
update enemy vision indicator

-vision indicator’s shape is now dynamically affected by obstacles
-player/enemy controller update function is now called in scene’s
didFinishUpdate fucntion

karamarimo commented:
やっと視野の表示できました。
なかなか難しかったです。

Commit:
fix GameScene memory leaks

-fixed enemy node not being freed after scene transition
-fixed player controller being created twice when app starts

Commit:
fix StartScene memory leaks

-fixed start scene and its button not being freed

karamarimo commented:
メモリリーク直りました。

karamarimo commented:
マージしてよろしいでしょうか

kameike commented:
反応遅うなってごめんねーいいのでは
